MDEV-24377: Accept comma separated addresses as --bind-address value

Paragoumba opened the pull request:
<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling this template <3

If you have any questions related to MariaDB or you just want to
hang out and meet other community members, please join us on
https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue
that seems to track this bug/feature, please add its number below.
-->
- [x] *The Jira issue number for this PR is: MDEV-24377*

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed, what was it looking like before
   the change and how it's looking with this patch applied
3. Do you think this patch might introduce side-effects in
   other parts of the server?
-->
## Description
Allow the option `--bind-address` to take multiple addresses separated by commas as a value. The server then listens on all the resolved addresses.

## How can this PR be tested?

This PR is not tested yet. It can be tested by ensuring it doesn't break existing tests and by adding tests starting the server with multiple valid and invalid values for the `--bind-address` option.

<!--
Tick one of the following boxes [x] to help us understand
if the base branch for the PR is correct
-->
## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

<!--
You might consider answering some questions like:
1. Does this affect the on-disk format used by MariaDB?
2. Does this change any behavior experienced by a user
   who upgrades from a version prior to this patch?
3. Would a user be able to start MariaDB on a datadir
   created prior to your fix?
-->
## Backward compatibility
This feature is backward compatible as all the values (single addresses without comma) accepted before implementing this feature are still accepted.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=2009) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=2009) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=2009) it.</sub>

LinuxJedi commented:
Hi @Paragoumba, could you please rebase this off 10.11 and add some test cases for this? We should then be able to review it for you.

LinuxJedi commented:
> Hi @LinuxJedi, sorry for the late response. As you asked I rebased the branch on the 10.11 branch. As fr the test cases I looked into it bu might need some help

Not a problem, I was on vacation until today too :)

For the test cases I suggest looking at `mysql-test/main/bind_address_resolution.*`, copying that and maybe use two addresses in the 127.* range or 127.0.0.1 and 0.0.0.0. Whilst this won't test everything it will at least test for a regression in the parsing and setup of the bind.

Paragoumba commented:
I created a unit test listening on both localhost (which should be resolved to 127.0.0.1 and ::1) and 192.168.0.1. The test doesn't pass as wsrep can't parse the argument. I tried to parse it ahead of time in `sys_vars.cc` with on ON_CHECK or ON_UPDATE function but couldn't make it work.

Paragoumba commented:
@LinuxJedi I am not sure how to make the test pass, how should I handle wsrep ?

LinuxJedi commented:
> @LinuxJedi I am not sure how to make the test pass, how should I handle wsrep ?

Hi!
I have a meeting with the core MariaDB engineering team tomorrow. I'll bring this up and see if I can get some help for you.

Paragoumba commented:
@LinuxJedi I modified wsrep so it only uses the first address. It shouldn't break anything and refactoring it to use multiple addresses seems to be a task too big for me. All the tests are passing now, including the one I added

Paragoumba commented:
@LinuxJedi I just reviewed the test I added and it appears it isn't testing properly my changes. A problem I am facing in order to refactor the test is that I can't add the addresses I want to the network interface, so the test fails if I use addresses that localhost doesn't resolve to by default.

Paragoumba commented:
bump @LinuxJedi

LinuxJedi reviewed (changes requested):
I think it is almost there. A couple of minor things. Also can you please add a test that does something like `SHOW VARIABLES LIKE 'bind_address'` to show that it has been processed at that side?

Paragoumba commented on sql/mysqld.cc:
```diff
@@ -2291,7 +2292,7 @@ static void activate_tcp_port(uint port,
                               bool is_extra_port= false)
 {
-  struct addrinfo *ai, *a;
+  struct addrinfo *ai, *a, *b = NULL, *head = NULL;
```
You are right, I deleted `b` which serves the same purpose as `a`. I believe requesting `head`'s deletion was a mistake on your part and kept it

Paragoumba commented on sql/mysqld.cc:
```diff
@@ -2312,16 +2313,50 @@ static void activate_tcp_port(uint port,
+      {
+        b= b->ai_next;
+      }
+      if (unlikely(error != 0))
```
Good catch, I move the error checking

grooverdan commented:
@vaintroub, the test cases here add a non-existing (at least in some environments) ip of 192.168.0.1 into the bind address tests. Linux obviously tolerates the address while its a 10049 error on Windows. With this it will be a [fatal error](https://buildbot.mariadb.org/#/builders/234/builds/12985/steps/9/logs/stdio). Is the last commit here that aims to make this test pass and only be fatal if no Windows addresses can be bound to acceptable?

grooverdan commented:
@brad0 do you know what the BSD behaviours are for binding to non-existent addresses?

brad0 commented:
> @brad0 do you know what the BSD behaviours are for binding to non-existent addresses?

I have this feeling like this is something I have seen added over the years, but I'll have to speak to others and get back to you.

grooverdan commented on sql/wsrep_utils.cc:
```diff
@@ -457,7 +457,13 @@ unsigned int wsrep_check_ip (const char* const addr, bool *is_ipv6)
 
-  int gai_ret= getaddrinfo(addr, NULL, &hints, &res);
+  char *end;
+  char address[FN_REFLEN];
```
FN_REFLEN was for path lengths so its quite overkill for an address length. `INET6_ADDRSTRLEN` is a good replacement.

grooverdan commented on sql/mysqld.cc:
```diff
@@ -135,6 +135,7 @@
 
 #define mysqld_charset &my_charset_latin1
 
+#define DELIM ','
```
seems used only once, so seems overkill to define it.

grooverdan reviewed (approved):
Overall pretty happy. Just waiting on @vaintroub to approve/reject the Windows handling of bind failures. The other small nits can be cleaned up in the final rebase/squash of commits.

Seems [FreeBSD](http://buildbot.askmonty.org/buildbot/builders/kvm-ports-freebsd130-amd64/builds/3983/steps/mtr/logs/stdio) doesn't like non-existent bind IPs either:
```
2022-10-27  3:25:51 0 [ERROR] Can't start server: Bind on TCP/IP port. Got error: 49: Can't assign requested address
2022-10-27  3:25:51 0 [ERROR] Do you already have another server running on port: 16040 ?
2022-10-27  3:25:51 0 [ERROR] Aborting
````
Errno 49 / [EADDRNOTAVAIL](https://www.unix.com/man-page/FreeBSD/2/errno/)

So just about to push these fixes too.

brad0 commented:
FreeBSD uses IP_BINDANY.


OpenBSD uses SO_BINDANY.

SO_BINDANY allows the socket to be bound to addresses which are not local to the machine, so it can be used to make a transparent proxy. Note that this option is limited to the superuser. In order to receive packets for these addresses, SO_BINDANY needs to be combined with matching outgoing [pf(4)](https://man.openbsd.org/pf.4) rules with the divert-reply parameter. For example, with the following rule the socket receives packets for 192.168.0.10 even if it is not a local address:

```pass out inet from 192.168.0.10 divert-reply```

Paragoumba commented:
I am very glad this has been merged. Thank you all for your help, especially @LinuxJedi and @grooverdan :grin:

grooverdan commented:
Thanks for bearing with a long contribution process. We're very happy to get your code and desired feature into MariaDB server for the use by everyone.
